### PR TITLE
adding coverage on project

### PR DESCRIPTION
**Pull request by balikasg**
This is a PR to add code coverage on the classification and ranking pipelines 
Since we are using 2 builds, we need to have a way to combine coverage from 2 runs. 
I followed the documentation (https://coverage.readthedocs.io/en/v4.5.x/cmd.html)
on how to user env variables and combine different outputs.

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -59,8 +59,12 @@ jobs:
           command: |
             . venv/bin/activate
             cd python
-            python3 -m pytest ml4ir/applications/classification/tests --junitxml=test_results/results.xml
+            COVERAGE_FILE=reports/.coverage.classif coverage run  --include "ml4ir/*" -m pytest ml4ir/applications/classification/tests/ --junitxml=test_results/results.xml
```
Comment (balikasg) [inline .circleci/config.yml:62]: We use an env variable `COVERAGE_FILE` to control the name of the file that results are written.

File: .circleci/config.yml [hunk 2]
```diff
@@ -59,8 +59,12 @@ jobs:
...
           no_output_timeout: 45m  # Time for running the tests before we timeout
+      - persist_to_workspace:
```
Comment (balikasg) [inline .circleci/config.yml:64]: This is used to share data between builds (https://circleci.com/docs/2.0/workflows/#using-workspaces-to-share-data-among-jobs)

File: .circleci/config.yml [hunk 3]
```diff
@@ -88,15 +92,56 @@ jobs:
...
+            COVERAGE_FILE=reports/.coverage.ranking coverage run --include "ml4ir/*" -m pytest ml4ir/applications/ranking/tests/ --junitxml=test_results/results.xml
           no_output_timeout: 45m  # Time for running the tests before we timeout
+      - persist_to_workspace:
+          root: python/reports
+          paths:
+            - .coverage.ranking
       - store_test_results:
           path: python/test_results
 
+  build_test_coverage:
```
Comment (balikasg) [inline .circleci/config.yml:104]: this is the new job that merges the 2 coverage files and publishes the report as an artifact.

File: .circleci/config.yml [hunk 4]
```diff
@@ -88,15 +92,56 @@ jobs:
...
+          path: htmlcov
+
 workflows:
   version: 2.1
   build_and_test:
     jobs:
       - build_test_jvm_integration
       - build_test_python_ranking
       - build_test_python_classification
+      - build_test_coverage:
```
Comment (balikasg) [inline .circleci/config.yml:144]: this last job should start after the 2 requirements complete.

File: .circleci/config.yml [hunk 5]
```diff
@@ -88,15 +92,56 @@ jobs:
...
+          root: python/reports
+          paths:
+            - .coverage.ranking
       - store_test_results:
           path: python/test_results
 
+  build_test_coverage:
+    docker:
+      - image: circleci/python:3.7.7
+    resource_class: large
```
Comment (ducouloa) [inline .circleci/config.yml:107]: Did you try with a standard Docker container?

I think, we have running costs starting with "large" or the one just below.

**Review by ducouloa — APPROVED** [review]
This looks good for me.

Do you know if we can imagine a check that ensure the _newly_ added code has a coverage > 85%?

If you add a small enough change you can left it untested I guess.

**Pull request merged**